[DX-993, TT-11639] Updated multiple middleware for OAS (#4012)
TykTechnologies/tyk-docs: tyk-docs/content/product-stack/tyk-gateway/middleware/request-header-tyk-classic.md

---
title: Using the Request Header Transform with Tyk Classic APIs
date: 2024-01-20
description: "Using the Request Header Transform middleware with Tyk Classic APIs"
tags: ["Request Header Transform", "middleware", "per-endpoint","per-API", "Tyk Classic"]
---

> ## Overview

Reviewer: Remove ##Overview heading so have page title, intro paragraph followed by section headings

> ## Overview

Author: ```suggestion
```

Tyk's [request header transform]({{< ref "transform-traffic/request-headers" >}}) middleware enables you to append or delete headers on requests to your API endpoints before they are passed to your upstream service.

There are two options for this:
 - API-level modification that is applied to all requests to the API
 - endpoint-level modification that is applied only to requests to a specific endpoint

{{< note success >}}
**Note**  

If both API-level and endpoint-level middleware are configured, the API-level transformation will be applied first.
{{< /note >}}

> When working with Tyk Classic APIs the transformation is configured in the Tyk Classic API Definition; this can be done manually within the `.json` file or from the API Designer in the Tyk Dashboard.

Reviewer: ```suggestion
When working with Tyk Classic APIs the transformation is configured in the Tyk Classic API Definition. This can be done manually within the `.json` file or from the API Designer in the Tyk Dashboard.
```

If you want to use dynamic data from context variables, you must [enable]({{< ref "context-variables#enabling-context-variables-for-use-with-tyk-classic-apis" >}}) context variables for the API to be able to access them from the request header transform middleware.

If you're using the newer Tyk OAS APIs, then check out the [Tyk OAS]({{< ref "product-stack/tyk-gateway/middleware/request-header-tyk-oas" >}}) page.

## Configuring the Request Header Transform in the Tyk Classic API Definition

> The API-level and endpoint-level request header transforms are configured in different sections of the API definition, though have a common configuration.

Reviewer: ```suggestion
The API-level and endpoint-level request header transforms are configured in different sections of the API definition and share a common configuration.
```

Author: I don't feel that "and" is the right word here - the first part of the phrase refers to a difference, the second to a similarity. I'll propose a re-phrasing.

Author: ```suggestion
The API-level and endpoint-level request header transforms have a common configuration but are configured in different sections of the API definition.
```

 - `X-Request-ID`
 - `X-User-ID`
 - `X-Secret`

and to remove one:
 - `Auth_Id` 

## Configuring the Request Header Transform in the API Designer
You can use the API Designer in the Tyk Dashboard to configure the request header transform middleware for your Tyk Classic API by following these steps.

#### API-level transform

Configuring the API-level request header transform middleware is very simple when using the Tyk Dashboard.

In the Endpoint Designer you should select the **Global Version Settings** and ensure that you have selected the **Request Headers** tab:

{{< img src="/img/2.10/global_settings_modify_headers.png" alt="Global version settings" >}}

Note that you must click **ADD** to add a header to the list (for appending or deletion).

#### Endpoint-level transform

#### Step 1: Add an endpoint for the path and select the Header Transform plugin
From the **Endpoint Designer** add an endpoint that matches the path for which you want to perform the transformation. Select the **Modify Headers** plugin.

#### Step 2: Configure the transform

> You must also set a method and a request pattern to match against. These patterns can contain wildcards in the form of any string bracketed by curly braces. These wildcards are so they are human readable and do not translate to variable names. Under the hood, a wildcard translates to the "match everything" regex of: `(.*)`.

Author: ```suggestion
You must also set a method and a request pattern (endpoint) to match against. These patterns can contain wildcards in the form of any string bracketed by curly braces. These wildcards are so they are human readable and do not translate to variable names. Under the hood, a wildcard translates to the "match everything" regex of: `(.*)`.
```
